crates/factor-outbound-http/src/wasi.rs

```diff
@@ -320,11 +326,17 @@ impl RequestSender {
             },
         );
 
+        // If we're limiting concurrent outbound requests, acquire a permit
+        let permit = match &self.concurrent_outbound_requests_semaphore {
+            Some(s) => s.acquire().await.ok(),
+            None => None,
+        };
         let resp = timeout(first_byte_timeout, resp)
             .await
             .map_err(|_| ErrorCode::ConnectionReadTimeout)?
             .map_err(hyper_legacy_request_error)?
             .map(|body| body.map_err(hyper_request_error).boxed());
+        drop(permit);
```

Comment:
We should either do as Lann suggested and limit by connection rather than request, or else attach `permit` to the response body so that it is only dropped when the body is dropped (which could be long after we've returned from this function).